EventFilter/HcalRawToDigi/plugins/HcalDigiToRawuHTR.cc

```diff
@@ -0,0 +1,297 @@
+  // HF (QIE8) precision data
+  // - - - - - - - - - - - - - - - - - - - - - - - - - - - 
+  // loop over each digi and allocate memory for each
+  for(HFDigiCollection::const_iterator qiedf=qie8hfdc.begin();qiedf!=qie8hfdc.end();qiedf++){
+    //HFDataFrame qiedf = static_cast<HFDataFrame>(qie8hfdc[j]);
+    HcalElectronicsId eid = qiedf->elecId();
+    int crateId = eid.crateId();
+    int slotId = eid.slot();
+    int uhtrIndex = (crateId&0xFF) | ((slotId&0xF)<<8) ; 
+
+    if( ! uhtrs.exist(uhtrIndex) ){
+      uhtrs.newUHTR( uhtrIndex );
+    }
+    uhtrs.addChannel(uhtrIndex,qiedf,_verbosity);
+  }
+
+  // - - - - - - - - - - - - - - - - - - - - - - - - - - - 
+  // HBHE (QIE8) precision data
+  // - - - - - - - - - - - - - - - - - - - - - - - - - - - 
+  // loop over each digi and allocate memory for each
+  for(HBHEDigiCollection::const_iterator qiedf=qie8hbhedc.begin();qiedf!=qie8hbhedc.end();qiedf++){
+    //HFDataFrame qiedf = static_cast<HFDataFrame>(qie8hbhedc[j]);
+    HcalElectronicsId eid = qiedf->elecId();
+    int crateId = eid.crateId();
+    int slotId = eid.slot();
+    int uhtrIndex = (crateId&0xFF) | ((slotId&0xF)<<8) ; 
+
+    if( ! uhtrs.exist(uhtrIndex) ){
+      uhtrs.newUHTR( uhtrIndex );
+    }
+    uhtrs.addChannel(uhtrIndex,qiedf,_verbosity);
+  }
+
+  // - - - - - - - - - - - - - - - - - - - - - - - - - - - 
+  // TP data
+  // - - - - - - - - - - - - - - - - - - - - - - - - - - - 
+  // loop over each digi and allocate memory for each
+
+  // --- I left off here ... need to specify and include the correct digi collection ---
+
+  for(HcalTrigPrimDigiCollection::const_iterator qiedf=qietpdc.begin();qiedf!=qietpdc.end();qiedf++){
+    //HFDataFrame qiedf = static_cast<HFDataFrame>(qie8hbhedc[j]);
+    HcalElectronicsId eid(qiedf->id().rawId());
+    
+    int crateId = eid.crateId();
+    int slotId = eid.slot();
+    int uhtrIndex = (crateId&0xFF) | ((slotId&0xF)<<8) ; 
+
+    if( ! uhtrs.exist(uhtrIndex) ){
+      uhtrs.newUHTR( uhtrIndex );
+    }
+    uhtrs.addChannel(uhtrIndex,qiedf,_verbosity);
+  }
+
+  // -----------------------------------------------------
+  // -----------------------------------------------------
+  // loop over each uHTR and format data
+  // -----------------------------------------------------
+  // -----------------------------------------------------
+  // loop over each uHTR and format data
+  int idxuhtr =-1;
+  for( UHTRMap::iterator uhtr = uhtrs.uhtrs.begin() ; uhtr != uhtrs.uhtrs.end() ; ++uhtr){
+
+    idxuhtr ++;
+   
+    uint64_t crateId = (uhtr->first)&0xFF;
+    uint64_t slotId =  (uhtr->first&0xF00)>>8;
+
+    uhtrs.finalizeHeadTail(&(uhtr->second),_verbosity);
+    int fedId = FEDNumbering::MINHCALuTCAFEDID + crateId;
+    if( fedMap.find(fedId) == fedMap.end() ){
+      /* QUESTION: where should the orbit number come from? */
+      fedMap[fedId] = new HCalFED(fedId,iEvent.id().event(),iEvent.orbitNumber(),iEvent.bunchCrossing());
```

Comment:
where is this deleted?
Consider using shared_ptr or unique_ptr instead of bare pointer